packages/components/src/components/Calculator/index.tsx

```diff
@@ -1,158 +1,237 @@
+    calculator.inputs.map((input) => [input.name, input.default ?? ""])
+  );
+}
+
+/**
+ * This function implements all state flow logic for calculator inputs.
+ */
+function useCalculator(
+  valueWithContext: SqCalculatorValueWithContext,
+  environment: Env
+) {
+  // useMemo here is important; valueWithContext.value is a getter that creates a new calculator value every time.
+  const calculator = useMemo(() => valueWithContext.value, [valueWithContext]);
+
+  const [cachedState, updateCachedState] =
+    useSavedCalculatorState(valueWithContext);
+
+  // Calculator input codes (Squiggle strings, selected options in <select>, etc.) are tracked by react-hook-form.
+  // Everything else is tracked in local state and backed up to ViewerContext.
+  const form = useForm<FormShape>({
+    defaultValues: cachedState?.formValues ?? getFormValues(calculator),
+  });
+
+  const [inputValues, setInputValues] = useState<InputValues>(
+    () => cachedState?.inputValues ?? {}
+  );
 
-import { Action, useViewerContext } from "../SquiggleViewer/ViewerProvider.js";
-import {
-  CalculatorAction,
-  CalculatorState,
-  calculatorReducer,
-  hasSameCalculator,
-  initialCalculatorState,
-} from "./calculatorReducer.js";
-import { CalculatorUI } from "./calculatorUI.js";
+  const hashString = useMemo(() => calculator.hashString, [calculator]);
+  const [prevHashString, setPrevHashString] = useState<string | undefined>(
+    cachedState?.hashString
+  );
+
+  // This effect does two things:
+  // - when the calculator is updated, it resets the form
+  // - also, on initial render and on calculator updates, it calculates all input values (which will trigger calculator result calculation)
+  // In both cases, it updates `prevHashString`, which tracks the calculator identity.
+  useEffect(() => {
+    if (prevHashString === hashString) {
+      return;
+    }
+
+    if (prevHashString) {
+      // Calculator identity has changed. Let's reset the form because there might be new inputs or new defaults.
+      // (In the future, we could do something more complicated here, for example check if any form fields were touched.)
+      form.reset(getFormValues(calculator));
+    }
+
+    // Calculator has updated (or this component just mounted). Let's take all input codes and run them.
+    const processAllFieldCodes = async () => {
+      const formValues = form.getValues();
+      const newInputValues: typeof inputValues = {};
+      for (const input of calculator.inputs) {
+        const name = input.name;
+        const fieldValue = formValues[name];
+        const inputValueResult = await runSquiggleCode(
+          fieldValueToCode(name, calculator, fieldValue),
+          environment
+        );
+        newInputValues[name] = inputValueResult;
+      }
+      // All values are updated simultaneously, to avoid too many rerenders.
+      setInputValues(newInputValues);
+    };
+    processAllFieldCodes();
+
+    setPrevHashString(hashString);
```

Comment:
I'd suggest moving this above the ``processAllFieldCodes`` function. Keep this useEffect() a bit separated between the "hashString" part and the processAllFieldCodes part.